codec/src/lib.rs

```diff
@@ -30,26 +30,29 @@
 //!     metadata: [u8; 11],
 //! }
 //!
-//! // Implement the Codec trait
-//! impl Codec for Point {
+//! // Implement the Encode trait
+//! impl Encode for Point {
+//!     fn len_encoded(&self) -> usize {
+//!       self.xy.len_encoded() + self.z.len_encoded() + self.metadata.len_encoded()
+//!     }
+//!
 //!     fn write(&self, buf: &mut impl BufMut) {
 //!         // Basic types can be written by inferring the type
 //!         self.xy.write(buf);
 //!         self.z.write(buf);
 //!         self.metadata.write(buf);
 //!     }
+//! }
 //!
-//!     fn read(buf: &mut impl Buf) -> Result<Self, Error> {
+//! // Implement the Decode trait
+//! impl Decode<()> for Point {
+//!     fn read(buf: &mut impl Buf, _: ()) -> Result<Self, Error> {
 //!         // Basic types can be inferred by the return type
-//!         let xy = <(u64, u64)>::read(buf)?;
-//!         let z = <Option<u32>>::read(buf)?;
-//!         let metadata = <[u8; 11]>::read(buf)?;
+//!         let xy = <(u64, u64)>::read(buf, ((), ()))?;
```

Comment:
I wonder if it makes sense to treat normal `read()` as having the empty cfg by default so that we don't need to  use these empty limits all over?

And then we could do `read_cfg()` if you want to provide a custom config? `read()` could be implemented as a default function in the trait?

Reply:
Or I suppose we could do `DecodeExt` that offers this functionality (and leave `read()` as unimplemented if a config is required).